pkg/controllers/management/globaldns/globaldnsprovider_catalog_launcher.go

```diff
@@ -0,0 +1,220 @@
+	}
+	return n
+}
+
+//sync is called periodically and on real updates
+func (n *ProviderCatalogLauncher) sync(key string, obj *v3.GlobalDNSProvider) (runtime.Object, error) {
+	if obj == nil || obj.DeletionTimestamp != nil {
+		return nil, nil
+	}
+	metaAccessor, err := meta.Accessor(obj)
+	if err != nil {
+		return nil, err
+	}
+	creatorID, ok := metaAccessor.GetAnnotations()[globalnamespacerbac.CreatorIDAnn]
+	if !ok {
+		return nil, fmt.Errorf("GlobalDNS %v has no creatorId annotation", metaAccessor.GetName())
+	}
+
+	if err := globalnamespacerbac.CreateRoleAndRoleBinding(globalnamespacerbac.GlobalDNSProviderResource, obj.Name,
+		obj.UID, obj.Spec.Members, creatorID, n.managementContext); err != nil {
+		return nil, err
+	}
+
+	//handle external-dns deployment
+	if obj.Spec.Route53ProviderConfig != nil {
+		return n.handleRoute53Provider(obj)
+	}
+
+	if obj.Spec.CloudflareProviderConfig != nil {
+		return n.handleCloudflareProvider(obj)
+	}
+
+	return nil, nil
+}
+
+func (n *ProviderCatalogLauncher) handleRoute53Provider(obj *v3.GlobalDNSProvider) (runtime.Object, error) {
+	rancherInstallUUID := settings.InstallUUID.Get()
+	//create external-dns route53 provider
+	answers := map[string]string{
+		"provider":      "aws",
+		"aws.zoneType":  "public",
+		"aws.accessKey": obj.Spec.Route53ProviderConfig.AccessKey,
+		"aws.secretKey": obj.Spec.Route53ProviderConfig.SecretKey,
+		"txtOwnerId":    rancherInstallUUID + "_" + obj.Name,
+		"rbac.create":   "true",
+		"logLevel":      "debug",
+		"policy":        "sync",
+	}
+
+	if obj.Spec.Route53ProviderConfig.RootDomain != "" {
+		answers["domainFilters[0]"] = obj.Spec.Route53ProviderConfig.RootDomain
+	}
+	return n.createUpdateExternalDNSApp(obj, answers)
+}
+
+func (n *ProviderCatalogLauncher) handleCloudflareProvider(obj *v3.GlobalDNSProvider) (runtime.Object, error) {
+	rancherInstallUUID := settings.InstallUUID.Get()
+	//create external-dns route53 provider
+	answers := map[string]string{
+		"provider":          "cloudflare",
+		"cloudflare.apiKey": obj.Spec.CloudflareProviderConfig.APIKey,
+		"cloudflare.email":  obj.Spec.CloudflareProviderConfig.APIEmail,
+		"txtOwnerId":        rancherInstallUUID + "_" + obj.Name,
+		"rbac.create":       "true",
+		"logLevel":          "debug",
+		"policy":            "sync",
+	}
+
+	if obj.Spec.CloudflareProviderConfig.RootDomain != "" {
+		answers["domainFilters[0]"] = obj.Spec.CloudflareProviderConfig.RootDomain
+	}
+
+	return n.createUpdateExternalDNSApp(obj, answers)
+}
+
+func (n *ProviderCatalogLauncher) createUpdateExternalDNSApp(obj *v3.GlobalDNSProvider, answers map[string]string) (runtime.Object, error) {
+	//check if provider already running for this GlobalDNSProvider.
+	existingApp, err := n.getProviderIfAlreadyRunning(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	if existingApp != nil {
+		//check if answers should be updated
+		if ifAnswersDiffer(existingApp.Spec.Answers, answers) {
```

Comment:
Maybe we should use `reflect.DeepEqual` instead of defining our own compare function.

Reply:
I think reflect.DeepEqual checks corresponding elements? We can't assume that answers array is sorted right?